src/handshake.ts

```diff
@@ -1,89 +1,125 @@
+import { Buffer } from "buffer";
+
 import { bytes, bytes32 } from "./@types/basic";
 import { NoiseSession, XXHandshake } from "./xx";
 import { KeyPair } from "./@types/libp2p";
-import { Buffer } from "buffer";
 import {
   createHandshakePayload,
   decodeMessageBuffer,
   encodeMessageBuffer,
   getHandshakePayload,
-  signPayload
+  logger, signEarlyDataPayload,
+  signPayload,
 } from "./utils";
 import { WrappedConnection } from "./noise";
 
-type handshakeType = "XX";
-
 export class Handshake {
-  private type: handshakeType;
-  private remotePublicKey: bytes;
+  public isInitiator: boolean;
+  public session: NoiseSession;
+
+  private libp2pPrivateKey: bytes;
+  private libp2pPublicKey: bytes;
   private prologue: bytes32;
   private staticKeys: KeyPair;
   private connection: WrappedConnection;
   private xx: XXHandshake;
 
   constructor(
-    type: handshakeType,
-    remotePublicKey: bytes,
+    isInitiator: boolean,
+    libp2pPrivateKey: bytes,
+    libp2pPublicKey: bytes,
     prologue: bytes32,
     staticKeys: KeyPair,
     connection: WrappedConnection,
+    handshake?: XXHandshake,
   ) {
-    this.type = type;
-    this.remotePublicKey = remotePublicKey;
+    this.isInitiator = isInitiator;
+    this.libp2pPrivateKey = libp2pPrivateKey;
+    this.libp2pPublicKey = libp2pPublicKey;
     this.prologue = prologue;
     this.staticKeys = staticKeys;
     this.connection = connection;
 
-    this.xx = new XXHandshake();
+    this.xx = handshake || new XXHandshake();
+    this.session = this.xx.initSession(this.isInitiator, this.prologue, this.staticKeys);
   }
 
   // stage 0
-  async propose(isInitiator: boolean, earlyData?: bytes): Promise<NoiseSession> {
-    const ns = await this.xx.initSession(isInitiator, this.prologue, this.staticKeys, this.remotePublicKey);
-
-    if (isInitiator) {
-      const signedPayload = signPayload(this.staticKeys.privateKey, getHandshakePayload(this.staticKeys.publicKey));
+  async propose(earlyData?: bytes): Promise<void> {
+    if (this.isInitiator) {
+      const signedPayload = signPayload(this.libp2pPrivateKey, getHandshakePayload(this.staticKeys.publicKey));
```

Comment:
add log that it's proposing

Reply:
generally try to log, if I'm initiator and when I start something or I received something and when I finish something

Reply:
agree, haven't finished with the logging